tokio/src/runtime/metrics/histogram/h2_histogram.rs

```diff
@@ -0,0 +1,479 @@
+    fn from_n_p(n: u32, p: u32, bucket_offset: usize) -> Self {
+        assert!(n >= p, "{n} (n) must be at least as large as {p} (p)");
+        let num_buckets = ((n - p + 1) * 1 << p) as usize - bucket_offset;
+        Self {
+            num_buckets,
+            p,
+            bucket_offset,
+        }
+    }
+
+    /// Creates a builder for [`LogHistogram`]
+    pub fn builder() -> LogHistogramBuilder {
+        LogHistogramBuilder::default()
+    }
+
+    /// The maximum value that can be stored before truncation in this histogram
+    pub fn max_value(&self) -> u64 {
+        let n = (self.num_buckets / (1 << self.p)) - 1 + self.p as usize;
+        (1_u64 << n) - 1
+    }
+
+    pub(crate) fn value_to_bucket(&self, value: u64) -> usize {
+        let index = bucket_index(value, self.p);
+        let offset_bucket = if index < self.bucket_offset as u64 {
+            0
+        } else {
+            index - self.bucket_offset as u64
+        };
+        let max = self.num_buckets - 1;
+        offset_bucket.min(max as u64) as usize
+    }
+
+    pub(crate) fn bucket_range(&self, bucket: usize) -> std::ops::Range<u64> {
+        let LogHistogram {
+            p,
+            bucket_offset,
+            num_buckets,
+        } = self;
+        let input_bucket = bucket;
+        let bucket = bucket + bucket_offset;
+        let range_start_0th_bucket = match input_bucket {
+            0 => Some(0_u64),
+            _ => None,
+        };
+        let range_end_last_bucket = match input_bucket {
+            n if n == num_buckets - 1 => Some(u64::MAX),
+            _ => None,
+        };
+        if bucket < 1 << p {
+            // The first set of buckets are all size 1
+            let bucket = bucket as u64;
+            range_start_0th_bucket.unwrap_or(bucket)..range_end_last_bucket.unwrap_or(bucket + 1)
+        } else {
+            // Determine which range of buckets we're in, then determine which bucket in the range it is
+            let bucket = bucket as u64;
+            let p = *p as u64;
+            let w = (bucket >> p) - 1;
+            let base_bucket = (w + 1) * (1_u64 << p);
+            let offset = bucket - base_bucket;
+            let s = 1_u64 << (w + p);
+            let start = s + (offset << w);
+            let end = s + ((offset + 1) << w);
+
+            range_start_0th_bucket.unwrap_or(start)..range_end_last_bucket.unwrap_or(end)
+        }
+    }
+}
+
+/// Configuration for a [`LogHistogram`]
+///
+/// The log-scaled histogram implements an H2 histogram where the first bucket covers
+/// the range from 0 to [`LogHistogramBuilder::min_value`] and the final bucket covers
+/// [`LogHistogramBuilder::max_value`] to infinity. The precision is bounded to the specified
+/// [`LogHistogramBuilder::max_error`]. Specifically, the precision is the next smallest value
+/// of `2^-p` such that it is smaller than the requested precision.
```

Comment:
This sentence has 2 `precision`s in it, should one of them be `maximum error`? Otherwise perhaps the first `precision` could be `actual precision` to contract `requested precision`?